### ISSUE TITLE
IDE0004 - Incorrect "Redundant cast" when not casting would cause sign extension

### ISSUE THREAD

**Issue by mklemarczyk** [dotnet/roslyn#4888]
You have to stop warning on "Redundant cast" for operations that can change number sign.

There are conditions for filter this warning:
1. If casting type use the same number of bytes or less.
2. If casting type change form from unsigned to signed and vice versa.

You really stop displaying warning in such that case:
```csharp
ulong somenumber = Math.Pow(2^63);
double factor = (long)somenumber; // Cast number to signed and store in double for future use as a factor
factor /= long.MaxValue;
```

Really do I need to write this?
```csharp
ulong somenumber = Math.Pow(2^63);
long temp = (long)somenumber;
double factor = temp;
factor /= long.MaxValue;
```

**Comment by mikedn** [comment]
The redundant cast diagnostic is indeed incorrect in this case and should be fixed but you may want to fix your example to compile for the sake of clarity. The first line is probably supposed to be

```C#
ulong somenumber = (ulong)Math.Pow(2, 63);
```

**Comment by Pilchie** [comment]
@DustinCampbell, you already fixed this.  Can you add a link to the original issue.

**Comment by DustinCampbell** [comment]
That's right. This is fixed by #4643 and no longer incorrectly reports an invalid cast:

![image](https://cloud.githubusercontent.com/assets/116161/9611963/ba016272-5097-11e5-8bb8-82eec5de0a7c.png)

**Comment by mklemarczyk** [comment]
When it will be fixed in Visual Studio?
By the way, line is a data from outside
```csharp
ulong somenumber = Math.Pow(2^63);
```

**Comment by Pilchie** [comment]
It will be included in the next update to Visual Studio 2015.
